### PR TITLE
fix: bitcoin zmq subscriptions are not re created after lost connection

### PR DESCRIPTION
**Pull request by cylewitruk**
## Description

Closes: #1038

## Changes

This PR changes the ZMQ implementation from the `zeromq` crate to [`bitcoincore-zmq`](https://crates.io/crates/bitcoincore-zmq), which under the hood uses the [`zmq`](https://crates.io/crates/zmq/0.10.0) crate. The reasons for this are outlined in #1038.

`bitcoincore-zmq` provides much of the boilerplate which we were doing ourselves, which was able to be removed, making for a much simpler implementation.

Note that this crate subscribes to all available topics, so what the signer receives is based on which topics are configured on the bitcoin node. For example, given the following configuration, both `hashblock` and `rawblock` messages will be received:
```
zmqpubhashblock=tcp://*:28332
zmqpubrawblock=tcp://*:28332
```

## Testing Information

All existing tests pass, and in addition, the `bitcoincore-zmq` crate runs [integration tests](https://github.com/antonilol/rust-bitcoincore-zmq/blob/HEAD/.github/workflows/integration_tests.yml#L19-L23) against the latest 5 versions of bitcoin-core which effectively increases our test coverage in this area.

I tested this fairly extensively in `devenv` and verified the re-establishment of the ZMQ socket connection. One thing to note if testing this in devenv is that you must only _stop_ the `bitcoin` container, not remove it, otherwise it will lose its chainstate and both the `stacks-blockchain` and `bitcoin-miner` containers will fail to resume.

## Security Considerations

The `zmq` crate which is used under the hood hasn't been updated in two years. As I mention in #1038, I see this as a minimal security risk as this code path is used _only_ for receiving block information from a Bitcoin node (which is expected to be "internal"), and the signer only connects out, inbound connections are not supported.

The zmq version appears to be 4.1.6, which is still a supported zmq version. Security advisories can be found [here](https://github.com/zeromq/libzmq/security) and [here](https://github.com/advisories?query=zeromq%2Flibzmq).

## Checklist:

- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by djordon — COMMENTED** [review]
Nice. This looks good, but would you mind updating the opening description with more detailed instructions for how to test the fix on devnet?

File: signer/src/bitcoin/zmq.rs [hunk 1]
```diff
@@ -18,193 +18,107 @@
...
+                    tracing::warn!(%fd, endpoint = event.source_url, "disconnected from ZeroMQ endpoint");
+                }
+                _ => {}
+            },
+            Ok(SocketMessage::Message(msg)) => match msg {
+                Message::Block(block, height) => {
+                    tracing::trace!(?block, %height, "received block");
+                }
+                Message::HashBlock(hash, height) => {
+                    tracing::trace!(?hash, %height, "received block hash");
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/bitcoin/zmq.rs]: [dd0468b](https://github.com/stacks-network/sbtc/pull/1062/commits/dd0468bbe68514ff982f73e38a5ec1ab151fa927)

File: signer/src/error.rs [hunk 2]
```diff
@@ -13,6 +13,14 @@ use crate::storage::model::SigHash;
 /// Top-level signer error
 #[derive(Debug, thiserror::Error)]
 pub enum Error {
+    /// An error occurred while attempting to connect to the Bitcoin Core ZMQ socket.
+    #[error("timed-out trying to connect to bitcoin-core ZMQ endpoint: {0}")]
+    BitcoinCoreZmqConnectTimeout(String),
+
+    /// An error was received from the Bitcoin Core ZMQ subscriber.
+    #[error("error from bitcoin-core ZMQ: {0}")]
+    BitcoinCoreZmq(#[from] bitcoincore_zmq::Error),
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/error.rs]: I'm fine changing it to #[source], it's just a couple of map_errors -- [dd0468b](https://github.com/stacks-network/sbtc/pull/1062/commits/dd0468bbe68514ff982f73e38a5ec1ab151fa927)

File: signer/src/bitcoin/zmq.rs [hunk 3]
```diff
@@ -18,193 +18,107 @@
...
-            Err(err) => ready(Some(Err(err))),
-        })
+        self.inspect(Self::inspect_message)
+            .filter_map(|msg| match msg {
+                Ok(SocketMessage::Message(msg)) => match msg {
+                    Message::HashBlock(hash, _) => ready(Some(Ok(hash))),
+                    _ => ready(None),
+                },
+                Err(err) => ready(Some(Err(err))),
+                Ok(_) => ready(None),
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/bitcoin/zmq.rs]: [dd0468b](https://github.com/stacks-network/sbtc/pull/1062/commits/dd0468bbe68514ff982f73e38a5ec1ab151fa927)

**Review by matteojug — APPROVED** [review]

**Pull request merged**